### PR TITLE
Find path to packages by looking for package.xml (fixes #175, fixes #181)

### LINKED ISSUE #175: Package database construction fails for unusual ROS_PACKAGE_PATH env vars
Usually, each element in the `ROS_PACKAGE_PATH` corresponds to a single package. ROSWire relies on this observation to find the location of each package and to parse its contents. However, in some cases, elements within the `ROS_PACKAGE_PATH` may correspond to a directory that contains several packages. ROSWire should accommodate for this case.

The best way to handle this issue is to do the following for each element in `ROS_PACKAGE_PATH`:
* check if the directory exists.
* check if the directory contains a `package.xml` file. If so, it's a package directory. If not, the element may correspond to a directory that hosts multiple ROS packages. In that case, recursively find all ROS packages within each of its subdirectories.

### LINKED ISSUE #181: Add find method to FileProxy
Should accept the following arguments (as a minimum):
* the absolute path of the directory that should be searched.
* the name of the file.

### PR DESCRIPTION
**Pull request by afsafzal**

### CODE REVIEW

File: src/roswire/definitions/package.py [hunk 1]
```diff
@@ -90,7 +90,22 @@ def paths(shell: ShellProxy) -> List[str]:
...
+        package_paths: List[str] = path_str.strip().split(':')
+        paths: List[str] = []
+        for path in package_paths:
+            print("P: {}".format(path))
+            code, _, _ = shell.execute('test -f {}/package.xml'.format(path))
+            print("code: {}".format(code))
+            if code != 0:
+                paths.append(path)
+            else:
+                code, dirs_str, _ = shell.execute('ls -d {}/*'.format(path))
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: Use `shlex.quote`.

File: src/roswire/definitions/package.py [hunk 2]
```diff
@@ -90,7 +90,22 @@ def paths(shell: ShellProxy) -> List[str]:
         code, path_str, duration = shell.execute('echo "${ROS_PACKAGE_PATH}"')
         if code != 0:
             raise Exception("unexpected error when fetching ROS_PACKAGE_PATH")
-        paths: List[str] = path_str.strip().split(':')
+        package_paths: List[str] = path_str.strip().split(':')
+        paths: List[str] = []
+        for path in package_paths:
+            print("P: {}".format(path))
+            code, _, _ = shell.execute('test -f {}/package.xml'.format(path))
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: `shlex.quote`.

File: src/roswire/definitions/package.py [hunk 3]
```diff
@@ -90,7 +90,22 @@ def paths(shell: ShellProxy) -> List[str]:
...
+            print("code: {}".format(code))
+            if code != 0:
+                paths.append(path)
+            else:
+                code, dirs_str, _ = shell.execute('ls -d {}/*'.format(path))
+                print("dirs_str: {}".format(dirs_str))
+                if code != 0:
+                    raise Exception("unexpected error when listing directories")
+                dirs: List[str] = path_str.strip().split('\n')
+                package_paths.extend(dirs)
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: It shouldn't be possible to modify a list whilst iterating over it. Maybe switch to using a Queue instead?

File: src/roswire/definitions/package.py (hunk 1 above)
Comment (ChrisTimperley) [reply src/roswire/definitions/package.py]: Or, better yet, use the FileProxy API to do this.

File: src/roswire/definitions/package.py (hunk 2 above)
Comment (ChrisTimperley) [reply src/roswire/definitions/package.py]: Use the FileProxy API.

File: src/roswire/definitions/package.py [hunk 4]
```diff
@@ -90,7 +91,25 @@ def paths(shell: ShellProxy) -> List[str]:
         code, path_str, duration = shell.execute('echo "${ROS_PACKAGE_PATH}"')
         if code != 0:
             raise Exception("unexpected error when fetching ROS_PACKAGE_PATH")
-        paths: List[str] = path_str.strip().split(':')
+        package_paths: List[str] = path_str.strip().split(':')
+        paths: List[str] = []
+        for path in package_paths:
+            cmd = f'test -f {shlex.quote(path)}/package.xml'
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: This is better, but it makes more sense to add `FileProxy` to the args for this function and to use its interface to perform this operation.

File: src/roswire/definitions/package.py [hunk 5]
```diff
@@ -90,7 +91,25 @@ def paths(shell: ShellProxy) -> List[str]:
...
+            code, _, _ = shell.execute(cmd)
+            if code == 0:
+                paths.append(path)
+            else:
+                cmd = f'ls -d {shlex.quote(path)}/*'
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: Same as above.

File: src/roswire/definitions/package.py [hunk 6]
```diff
@@ -90,7 +91,25 @@ def paths(shell: ShellProxy) -> List[str]:
...
+                code, dirs_str, _ = shell.execute(cmd)
+                if code != 0:
+                    # Not a directory
+                    continue
+                dirs: List[str] = dirs_str.strip().split('\n')
+                for d in dirs:
+                    cmd = f'test -f {shlex.quote(d)}/package.xml'
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: Same as above.

File: src/roswire/definitions/package.py [hunk 7]
```diff
@@ -83,7 +83,7 @@ def to_dict(self) -> Dict[str, Any]:
 
 class PackageDatabase(Mapping[str, Package]):
     @staticmethod
-    def paths(shell: ShellProxy) -> List[str]:
+    def paths(shell: ShellProxy, file_proxy: FileProxy) -> List[str]:
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: Just `files` is okay.

File: src/roswire/definitions/package.py [hunk 8]
```diff
@@ -82,15 +83,28 @@ def to_dict(self) -> Dict[str, Any]:
...
+            if file_proxy.isfile(f'{path}/package.xml'):
+                paths.append(path)
+            else:
+                try:
+                    dirs: List[str] = file_proxy.listdir(path, absolute=True)
+                except OSError:
+                    # Not a directory
+                    continue
+                for d in dirs:
+                    if file_proxy.isfile(f'{d}/package.xml'):
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: This doesn't work for cases where `package.xml` is nested more than one subdirectory deep. Using a queue is probably better.

File: src/roswire/definitions/package.py [hunk 9]
```diff
@@ -82,15 +83,28 @@ def to_dict(self) -> Dict[str, Any]:
...
             raise Exception("unexpected error when fetching ROS_PACKAGE_PATH")
-        paths: List[str] = path_str.strip().split(':')
+        package_paths: List[str] = path_str.strip().split(':')
+        paths: List[str] = []
+        for path in package_paths:
+            if file_proxy.isfile(f'{path}/package.xml'):
+                paths.append(path)
+            else:
+                try:
+                    dirs: List[str] = file_proxy.listdir(path, absolute=True)
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: You could just extend a queue with all of the subdirectories at this point.

File: src/roswire/definitions/package.py (hunk 8 above)
Comment (afsafzal) [reply src/roswire/definitions/package.py]: I tried that. It takes a very long time to traverse all subdirectories. Another way is to just use `find` to list all `package.xml` files. Do you think that's a better solution?

File: src/roswire/definitions/package.py (hunk 9 above)
Comment (afsafzal) [reply src/roswire/definitions/package.py]: The comment above.

File: src/roswire/definitions/package.py (hunk 8 above)
Comment (ChrisTimperley) [reply src/roswire/definitions/package.py]: That sounds good. It would be great if we could add an interface for `find` to `FileProxy`.

Comment (ChrisTimperley) [reply src/roswire/definitions/package.py]: I've created a new issue: https://github.com/ChrisTimperley/roswire/issues/181

File: src/roswire/proxy/file.py [hunk 10]
```diff
@@ -186,6 +186,28 @@ def islink(self, path: str) -> bool:
...
 
+    def find(self,
+             path: str,
+             filename: str
+             ) -> List[str]:
+        """
+        Returns a list of files that match the provided filename
+        in a given directory recursively.
+
+        Parameters:
```
Comment (ChrisTimperley) [inline src/roswire/proxy/file.py]: This is good, but not quite the right format: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

File: src/roswire/definitions/package.py [hunk 11]
```diff
@@ -94,17 +94,13 @@ def paths(shell: ShellProxy, file_proxy: FileProxy) -> List[str]:
...
-                    dirs: List[str] = file_proxy.listdir(path, absolute=True)
-                except OSError:
-                    # Not a directory
-                    continue
-                for d in dirs:
-                    if file_proxy.isfile(f'{d}/package.xml'):
-                        paths.append(d)
+            try:
+                all_packages = files.find(path, 'package.xml')
+            except OSError:
```
Comment (ChrisTimperley) [inline src/roswire/definitions/package.py]: https://docs.python.org/3/library/contextlib.html#contextlib.suppress

File: src/roswire/proxy/file.py [hunk 12]
```diff
@@ -186,6 +186,28 @@ def islink(self, path: str) -> bool:
...
+            filename: the name of the file to match.
+
+        Raises:
+            OSError: if find operation is not successful.
+        """
+        cmd = f'find {shlex.quote(path)} -name {shlex.quote(filename)}'
+        code, output, duration = self.__shell.execute(cmd)
+        if code != 0:
+            raise OSError(f"Find operation was unsuccessful on {path}")
+        paths: List[str] = output.replace('\r', '').split('\n')
```
Comment (ChrisTimperley) [inline src/roswire/proxy/file.py]: Will this return the absolute file path? It might be worth writing a test for this method.

Comment (afsafzal) [reply src/roswire/proxy/file.py]: I think it returns `path`+subdirectories. So if `path` is absolute, the returned paths are absolute too.

File: src/roswire/definitions/package.py (hunk 11 above)
Comment (afsafzal) [reply src/roswire/definitions/package.py]: But I don't want to suppress it. I want to continue the for loop if there's an exception.

File: src/roswire/proxy/file.py (hunk 10 above)
Comment (afsafzal) [reply src/roswire/proxy/file.py]: Is this format different than what you've used in other functions of this class?

File: src/roswire/proxy/file.py (hunk 12 above)
Comment (afsafzal) [reply src/roswire/proxy/file.py]: I'll add a test.

File: src/roswire/definitions/package.py (hunk 11 above)
Comment (ChrisTimperley) [reply src/roswire/definitions/package.py]: Ah. Good point.

File: src/roswire/proxy/file.py (hunk 10 above)
Comment (ChrisTimperley) [reply src/roswire/proxy/file.py]: Apparently I've messed up the documentation in this module. Don't worry about it. I'll create an issue to look at the documentation later.

**Pull request merged**